### ISSUE TITLE
Support `AbstractFilePath`

### ISSUE THREAD

**Issue by hasufell** [haskell/directory#138]
## Motivation

The next filepath release will add support for a new API, which fixes subtle encoding issues and improves cross platform code and memory residence.

You can read about the state of the newly added API here: https://github.com/haskellfoundation/tech-proposals/issues/35

Release candidate with haddock is here: https://hackage.haskell.org/package/filepath-2.0.0.3/candidate

Demonstration about unsoundness of base with exotic encodings and how the new API fixes them is here: https://gist.github.com/hasufell/c600d318bdbe010a7841cc351c835f92

## Migrations

Migrations can happen once the `filepath`, `unix` and `Win32` packages are updated and in sync. A migration would usually involve using the new types from [System.AbstractFilePath](https://hackage.haskell.org/package/filepath-2.0.0.3/candidate/docs/System-AbstractFilePath.html) and the new API variants from [unix](https://github.com/haskell/unix/pull/202)/[Win32](https://github.com/haskell/win32/pull/198).

When writing low-level cross platform code manually (shouldn't generally be necessary), the usual strategy is this:

```hs
#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
import qualified System.AbstractFilePath.Data.ByteString.Short.Word16 as SBS
import qualified System.AbstractFilePath.Windows as PFP
#else
import qualified System.AbstractFilePath.Data.ByteString.Short as SBS
import qualified System.AbstractFilePath.Posix as PFP
#endif

crossPlatformFunction :: AbstractFilePath -> IO ()
#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
crossPlatformFunction (OsString pfp@(WS ba)) = do
    -- use filepath functions for windows specific operating system strings
    let ext = PFP.takeExtension pfp
    -- operate directly on the underlying bytestring (which is a wide character bytestring, so uses Word16)
    let foo = SBS.takeWhile
    ...
#else
crossPlatformFunction (OsString pfp@(PS ba)) = do
    -- use filepath functions for posix specific operating system strings
    let ext = PFP.takeExtension pfp
    -- operate directly on the underlying bytestring (which is just Word8 bytestring)
    let foo = SBS.takeWhile
    ...
#endif
```

Platform specific code can be written using [`PosixFilePath`](https://hackage.haskell.org/package/filepath-2.0.0.3/candidate/docs/System-AbstractFilePath-Posix.html)/[`WindowsFilePath`](https://hackage.haskell.org/package/filepath-2.0.0.3/candidate/docs/System-AbstractFilePath-Windows.html) types.

If you have further questions, please let me know. I'm going to write a blog post outlining the affairs and more in-depth intro and migration strategies close after the release. This is a heads up.

----

Related PR: https://github.com/haskell/directory/pull/136

**Comment by Rufflewind** [comment]
The bulk of the work has been merged (https://github.com/haskell/directory/pull/136).

There are still some outsanding TODOs, so leaving this issue open for the time being.

**Comment by Rufflewind** [comment]
Does not look like the TODO will be resolved any time soon, so moving that into #182.

The rest of this issue is done so marking as closed.
